mlir/include/mlir/Dialect/DLTI/DLTIAttrs.td

```diff
@@ -0,0 +1,207 @@
+            "GPU": #dlti.target_device_spec<
+                    #dlti.dl_entry<"dlti.max_vector_op_width", 64 : ui32>>,
+            "XPU": #dlti.target_device_spec<
+                    #dlti.dl_entry<"dlti.max_vector_op_width", 4096 : ui32>>>
+  }];
+  let parameters = (ins
+    ArrayRefParameter<"DeviceIDTargetDeviceSpecPair", "">:$entries
+  );
+  let mnemonic = "target_system_spec";
+  let genVerifyDecl = 1;
+  let assemblyFormat = "`<` $entries `>`";
+  let extraClassDeclaration = [{
+    /// Return the device specification that matches the given device ID
+    std::optional<TargetDeviceSpecInterface>
+    getDeviceSpecForDeviceID(
+      TargetSystemSpecInterface::DeviceID deviceID);
+  }];
+  let extraClassDefinition = [{
+    std::optional<TargetDeviceSpecInterface>
+    $cppClass::getDeviceSpecForDeviceID(
+        TargetSystemSpecInterface::DeviceID deviceID) {
+      for (const auto& entry : getEntries()) {
+        if (entry.first == deviceID)
+          return entry.second;
+      }
+      return std::nullopt;
+    }
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// TargetDeviceSpecAttr
+//===----------------------------------------------------------------------===//
+
+def TargetDeviceSpecTrait
+    : NativeAttrTrait<"TargetDeviceSpecInterface::Trait"> {
+  let cppNamespace = "::mlir";
+}
+
+def DLTI_TargetDeviceSpecAttr :
+    DLTIAttr<"TargetDeviceSpec", [TargetDeviceSpecTrait]> {
+  let summary = [{
+    An attribute to represent target device specification.
+  }];
+  let description = [{
+    Each device specification describes a single device and its
+    hardware properties. Each device specification can contain any number
+    of optional hardware properties (e.g., max_vector_op_width below).
+
+    Example:
+      #dlti.target_device_spec<
+        #dlti.dl_entry<"dlti.max_vector_op_width", 64 : ui32>>
+  }];
+  let parameters = (ins
+    ArrayRefParameter<"DataLayoutEntryInterface", "">:$entries
+  );
+  let mnemonic = "target_device_spec";
+  let genVerifyDecl = 1;
+  let assemblyFormat = "`<` $entries `>`";
+  let extraClassDeclaration = [{
+    /// Returns max vector op width identifier.
+    StringAttr getMaxVectorOpWidthIdentifier();
+
+    /// Returns L1 cache size identifier
+    StringAttr getL1CacheSizeInBytesIdentifier();
+
+    /// Returns the interface spec for max vector op width
+    /// Since max vector op width is an optional property, this function will
+    /// return a valid spec if the property is defined, otherwise it
+    /// will return an empty spec.
+    DataLayoutEntryInterface getSpecForMaxVectorOpWidth();
+
+    /// Returns the interface spec for L1 cache size
+    /// Since L1 cache size is an optional property, this function will
+    /// return a valid spec if the property is defined, otherwise it
+    /// will return an empty spec.
+    DataLayoutEntryInterface getSpecForL1CacheSizeInBytes();
```

Comment:
Still have specialized APIs here? (L1 and vector width)

Reply:
Removed these APIs now.